src/operator/upsert.rs

```diff
@@ -1,140 +1,206 @@
 use crate::{
-    algebra::{HasOne, HasZero},
+    algebra::{AddAssignByRef, HasOne, HasZero, PartialOrder, ZRingValue},
     circuit::{
         operator_traits::{BinaryOperator, Operator},
-        OwnershipPreference, Scope,
+        ExportId, ExportStream, OwnershipPreference, Scope,
     },
-    trace::{cursor::Cursor, Batch, Trace},
+    operator::trace::{DelayedTraceId, TraceAppend, TraceId, Z1Trace},
+    trace::{
+        consolidation::consolidate, cursor::Cursor, spine_fueled::Spine, Batch, Builder, Trace,
+        TraceReader,
+    },
+    utils::VecExt,
+    Circuit, Stream, Timestamp,
 };
-use std::{borrow::Cow, marker::PhantomData, mem::swap, ops::Neg};
+use size_of::SizeOf;
+use std::{borrow::Cow, marker::PhantomData, ops::Neg};
+
+impl<P, K, V> Stream<Circuit<P>, Vec<(K, Option<V>)>>
+where
+    P: Clone + 'static,
+{
+    /// Convert a stream of upserts into a stream of updates.
+    ///
+    /// The input stream carries changes to a key/value map in the form of
+    /// _upserts_.  An upsert assigns a new value to a key (or `None` to
+    /// remove the key fro the map) without explicitly removing the old
+    /// value, if any.  Upserts are produced by some operators
+    /// or arrive from external data sources via
+    /// [`UpsertHandle`](`crate::UpsertHandle`)s.  The operator converts upserts
+    /// into batches of updates, which is the input format of most DBSP
+    /// operators.
+    ///
+    /// The operator assumes that the input vector is sorted by key and contains
+    /// exactly one value per key.
+    ///
+    /// This is a stateful operator that internaly maintains the trace of the
+    /// collection.
+    // TODO: Derive TS from circuit.
+    pub fn upsert<TS, B>(&self) -> Stream<Circuit<P>, B>
+    where
+        K: Ord + Clone + Eq + SizeOf + 'static,
+        V: Ord + Clone + Ord + SizeOf + 'static,
+        B::R: ZRingValue + SizeOf,
+        TS: Timestamp + SizeOf,
+        B: Batch<Key = K, Val = V, Time = ()> + 'static,
+    {
+        let circuit = self.circuit();
+
+        // We build the following circuit to implement the upsert semantics.
```

Comment:
can distinct be implemented using upsert too?

Reply:
There are two ways to implement `distinct`: (1) assemble output into a trace and use upsert to determine values to retract, (2) compute retractions dynamically at the same time we scan the _input_ trace.  The second option can be more memory-efficient as it doesn't require output trace, and that's what we implement today, hence no need for upsert.

Reply:
but distinct needs to keep the output trace somewhere too, right?

Reply:
It doesn't if you compute the value to retract from the input trace.

Reply:
Is the input trace the integral of the input?
distinct needs to keep the integral around.

Reply:
A trace is the history of the input.  It can be used to compute the integral for any timestamp, current or past.

Reply:
We need a glossary. I think of a trace as Frank's datatype.

Reply:
Yes, it is.